exp show: Add --keep and --drop (#3112)
iterative/dvc.org: content/docs/command-reference/exp/show.md

to standard output.

By default, the printed experiments table will include columns for all metrics

> and params from the entire project. The `--param-deps`, `--drop`, `--keep` and

Reviewer: Unrelated to this PR, but why call out `--param-deps` first here? It seems like a less common option. Could `--only-changed` be called out first instead since it's often a useful option?

other [options](#options) can determine which ones should be displayed.

> experiments in the table are first grouped (by parent commit). They are then

Reviewer: Why this change?

Author: messed up with search and replace

- `--param-deps` - include only parameters that are stage dependencies.

- `--only-changed` - show only parameters and metrics with values that vary
  across experiments. Note that this option takes precedence over

>   `--include-params` and `--include-metrics`, for example given

Reviewer: Agreed. I see it in the examples, but probably good to be explicit here as well.

> - `--drop <list>` - remove the columns matching the specified regex patterns.

Reviewer: Does this accept a comma-separated list or just a single regex? Could you explain this a bit more and how to use in the basic case where regex isn't needed? Probably not all users will be familiar with regex.

Author: Does it make sense to keep supporting a list (comma separated `--drop 'x,y'` or multiple `--drop x --drop y`) now that it supports regex?

I understand that some users might not be familiar with regex but, if we show an example of how `--drop 'x|y'` would be equivalent as passing multiple values, I'm no longer sure if the lack of regex knowledge would be a problem at all.

Reviewer 2: Could there be more than one `--drop` or `--keep` in a command? Like `dvc exp show --drop this --drop that`?

Author: Updated core to accept single value only and added example for `x|y`

Author: > Could there be more than one `--drop` or `--keep` in a command? Like `dvc exp show --drop this --drop that`?

No. I added an example on how `this|that` could achieve the same

- `--keep <regex>` - prevent the columns matching the specified [regex][regex]

>   to be removed by any of the other options.

Reviewer 3: I hope we can introduce them without the concept of regex at first, and then mention separately any regex is accepted.

Author: I'm not sure what term to use here instead of regex.

Reviewer: "Pattern"? "Name"? Or simply "remove the matching columns."

Author: ```suggestion
- `--drop <regex>` - remove the matching columns.

- `--keep <regex>` - prevent the matching columns to be removed by any of the other options.
```

│ master                  │ May 29, 2021 │  0.60405 │  0.9608 │ 0.2           │ 20170428     │ 3000                   │ 2                │ 20170428   │ 100         │ 64              │
│ ├── d384680 [exp-bc055] │ 08:03 PM     │  0.51799 │ 0.92333 │ 0.2           │ 20170428     │ 500                    │ 2                │ 20170428   │ 100         │ 64              │
│ ├── 6b338f8 [exp-3315b] │ 08:03 PM     │  0.58589 │   0.945 │ 0.2           │ 20170428     │ 2000                   │ 2                │ 20170428   │ 100         │ 64              │
│ └── d7fdde2 [exp-1b262] │ 08:03 PM     │  0.56447 │ 0.94713 │ 0.2           │ 20170428     │ 1500                   │ 2                │ 20170428   │ 100         │ 64              │
└─────────────────────────┴──────────────┴──────────┴─────────┴───────────────┴──────────────┴────────────────────────┴──────────────────┴────────────┴─────────────┴─────────────────┘
```

> You can exit this screen with `Q`, typically.

As a quick way of reducing noise, `--only-changed` will drop any column with

> values that do not change across the shown experiments:

Reviewer 2: ~~shown~~

```dvc
$ dvc exp show --only-changed
```

```dvctable
┏━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ neutral:**Experiment**              ┃ neutral:**Created**      ┃ metric:**avg_prec** ┃ metric:**roc_auc** ┃ param:**featurize.max_features** ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│ workspace               │ -            │  0.60405 │  0.9608 │ 3000                   │
│ master                  │ May 29, 2021 │  0.60405 │  0.9608 │ 3000                   │
│ ├── d7fdde2 [exp-1b262] │ 08:03 PM     │  0.56447 │ 0.94713 │ 1500                   │
│ ├── 6b338f8 [exp-3315b] │ 08:03 PM     │  0.58589 │   0.945 │ 2000                   │
│ └── d384680 [exp-bc055] │ 08:03 PM     │  0.51799 │ 0.92333 │ 500                    │
└─────────────────────────┴──────────────┴──────────┴─────────┴────────────────────────┘
```

You can also use `--drop` to filter columns matching specific regex patterns:

```dvc

> $ dvc exp show --drop 'prepare'

Reviewer: Very minor, but maybe exclude quotes when doing simple exact column matches so they don't appear to be required.

┡━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ workspace               │ -            │  0.60405 │  0.9608 │ 3000                   │ 2                │ 20170428   │ 100         │ 64              │
│ master                  │ May 29, 2021 │  0.60405 │  0.9608 │ 3000                   │ 2                │ 20170428   │ 100         │ 64              │
│ ├── 6b338f8 [exp-3315b] │ 08:03 PM     │  0.58589 │   0.945 │ 2000                   │ 2                │ 20170428   │ 100         │ 64              │
│ ├── d384680 [exp-bc055] │ 08:03 PM     │  0.51799 │ 0.92333 │ 500                    │ 2                │ 20170428   │ 100         │ 64              │
│ └── d7fdde2 [exp-1b262] │ 08:03 PM     │  0.56447 │ 0.94713 │ 1500                   │ 2                │ 20170428   │ 100         │ 64              │
└─────────────────────────┴──────────────┴──────────┴─────────┴────────────────────────┴──────────────────┴────────────┴─────────────┴─────────────────┘
```

Combine `--only-changed` with `--drop` and/or `--keep`:

```dvc

> $ dvc exp show --only-changed --drop 'neutral:**Created**' --keep 'train.(?!seed)'

Reviewer: ```suggestion
$ dvc exp show --only-changed --drop Created --keep 'train.(?!seed)'
```

Was it intentional to include this syntax formatting? It's showing up as text in the deployment:

<img width="699" alt="Screen Shot 2021-12-20 at 10 38 33 AM" src="https://user-images.githubusercontent.com/2308172/146793335-35c9befe-3f21-46dd-b246-e0d99efebd25.png">

Author: It was not intentional

```

```dvctable
┏━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ neutral:**Experiment**              ┃ metric:**avg_prec** ┃ metric:**roc_auc** ┃ param:**featurize.max_features** ┃ param:**train.n_est** ┃ param:**train.min_split** ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ workspace               │  0.60405 │  0.9608 │ 3000                   │ 100         │ 64              │
│ master                  │  0.60405 │  0.9608 │ 3000                   │ 100         │ 64              │
│ ├── d384680 [exp-bc055] │  0.51799 │ 0.92333 │ 500                    │ 100         │ 64              │
│ ├── 6b338f8 [exp-3315b] │  0.58589 │   0.945 │ 2000                   │ 100         │ 64              │
│ └── d7fdde2 [exp-1b262] │  0.56447 │ 0.94713 │ 1500                   │ 100         │ 64              │
└─────────────────────────┴──────────┴─────────┴────────────────────────┴─────────────┴─────────────────┘
```

> Sort experiments by the `auc` metric, in ascending order:

Reviewer: Hm, is it better to specify `ascending order` here if the command doesn't use `--sort-order`? It shows default behavior but doesn't show use of `--sort-order` like it did before.

Author: Should I replace with `descending` order ? 

I think showing an option that it's not needed might be counterproductive.

Reviewer: Agreed. I'm fine with whatever you want to do here.

```dvc

> $ dvc exp show --only-changed --sort-by=metric:**roc_auc**

Reviewer: ```suggestion
$ dvc exp show --only-changed --sort-by=roc_auc
```
Same as above.

Reviewer: It probably is good to bold `--sort-by=roc_auc` to highlight the relevant part of the command.

Author: I don't think it's possible to do inside a code block.

Author: cc @iterative/websites ?

Reviewer 4: I don't think it's possible, but we could look into adding this as a feature if it's something we want.

Reviewer: Not a big deal. I guess it's also possible to do with [html](https://meta.stackexchange.com/questions/183610/how-to-combine-bold-and-code-sample-in-markdown), but let's not worry about it now.

> can find the actual source code.

Let's say we have run 3 experiments in our project. The basic usage shows the
workspace (Git working tree) and experiments derived from `HEAD` (`master`

> branch in this case), and all of their metrics and params (scroll right to see

Reviewer 3: Great job updating the example @daavoo, thanks. It's basically a guide so I see why you're not sure what to do in the UG. I think you shouldn't worry about the guide except minimal updates to reflect this change. We have not finished moving some of this info from the ref to the guide anyway so we'll have to revisit this.

```

> You can exit this screen with `Q`, typically.

As a quick way of reducing noise, `--only-changed` will drop any column with
values that do not change across experiments:

```dvc
$ dvc exp show --only-changed
```

```dvctable
┏━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ neutral:**Experiment**              ┃ neutral:**Created**      ┃ metric:**avg_prec** ┃ metric:**roc_auc** ┃ param:**featurize.max_features** ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│ workspace               │ -            │  0.60405 │  0.9608 │ 3000                   │
│ master                  │ May 29, 2021 │  0.60405 │  0.9608 │ 3000                   │
│ ├── d7fdde2 [exp-1b262] │ 08:03 PM     │  0.56447 │ 0.94713 │ 1500                   │
│ ├── 6b338f8 [exp-3315b] │ 08:03 PM     │  0.58589 │   0.945 │ 2000                   │
│ └── d384680 [exp-bc055] │ 08:03 PM     │  0.51799 │ 0.92333 │ 500                    │
└─────────────────────────┴──────────────┴──────────┴─────────┴────────────────────────┘
```

> You can also use `--drop` to filter columns matching a specific [regex][regex]:

Reviewer 3: Again the concept of regex is too technical. I wouldn't mention it down here:
```suggestion
You can also use `--drop` to filter specific columns:
```
> Unless you actually put a (clear) regex in the example

┡━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ workspace               │ -            │  0.60405 │  0.9608 │ 3000                   │ 2                │ 20170428   │ 100         │ 64              │
│ master                  │ May 29, 2021 │  0.60405 │  0.9608 │ 3000                   │ 2                │ 20170428   │ 100         │ 64              │
│ ├── 6b338f8 [exp-3315b] │ 08:03 PM     │  0.58589 │   0.945 │ 2000                   │ 2                │ 20170428   │ 100         │ 64              │
│ ├── d384680 [exp-bc055] │ 08:03 PM     │  0.51799 │ 0.92333 │ 500                    │ 2                │ 20170428   │ 100         │ 64              │
│ └── d7fdde2 [exp-1b262] │ 08:03 PM     │  0.56447 │ 0.94713 │ 1500                   │ 2                │ 20170428   │ 100         │ 64              │
└─────────────────────────┴──────────────┴──────────┴─────────┴────────────────────────┴──────────────────┴────────────┴─────────────┴─────────────────┘
```

You can use `|` in order to remove a list of columns:

```dvc

> $ dvc exp show --drop 'avg_prec|train.min_split'

Reviewer 3: Hmmmm. Do we recommend this over several `--drop`s? Not sure. Feels like we're documenting regular expressions here. We could skip this one (but maybe add another `--drop` to the previous one).

Author: I've removed support for multiple `--drop` in core.

I've considered that joining using `|`  was an easy to teach replacement for either multiple `--drop` or comma-separated lists.

Do you consider that we should still support multiple values?

Reviewer 2: > tend to think contrary to @iesahin. Examples in the ref. should serve as quick ways for users to grasp how the tool does its job and cover some main use cases. In the UG it's unclear we want separate example sections, as we can embed the examples throughout it. I believe @shcheklein also favors this view.

I believe REF examples should be like [tldr](https://tldr.sh/) pages, and without much explanation. People (like me) usually refer to REF to find the exact option quickly.

For UG, I tend to have more explanations, describing "why" and "how" rather verbosely. These may be separated from the UG, though I believe having UG around workflows (rather than commands) might make more sense. In this case, we'll have larger examples.

Reviewer: How about something like, "You can use regex to match columns. For example, to remove multiple columns:"

┡━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━┩
│ workspace               │ -            │  0.9608 │ 0.2           │ 20170428     │ 3000                   │ 2                │ 20170428   │ 100         │
│ master                  │ May 29, 2021 │  0.9608 │ 0.2           │ 20170428     │ 3000                   │ 2                │ 20170428   │ 100         │
│ ├── d384680 [exp-bc055] │ Dec 17, 2021 │ 0.92333 │ 0.2           │ 20170428     │ 500                    │ 2                │ 20170428   │ 100         │
│ ├── d7fdde2 [exp-1b262] │ Dec 17, 2021 │ 0.94713 │ 0.2           │ 20170428     │ 1500                   │ 2                │ 20170428   │ 100         │
│ └── 6b338f8 [exp-3315b] │ Dec 17, 2021 │   0.945 │ 0.2           │ 20170428     │ 2000                   │ 2                │ 20170428   │ 100         │
└─────────────────────────┴──────────────┴─────────┴───────────────┴──────────────┴────────────────────────┴──────────────────┴────────────┴─────────────┘
```

> Combine `--only-changed` with `--drop` and/or `--keep`:

Reviewer 3: ```suggestion
If combined `--only-changed` has the least priority, `--drop` comes next,
and `--keep` has the last word:
```

Author: Do you think we should still mention priorities on the `Options` section of this would be enough ?

```dvc
$ dvc exp show --only-changed --drop Created --keep 'train.(?!seed)'
```

```dvctable
┏━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ neutral:**Experiment**              ┃ metric:**avg_prec** ┃ metric:**roc_auc** ┃ param:**featurize.max_features** ┃ param:**train.n_est** ┃ param:**train.min_split** ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ workspace               │  0.60405 │  0.9608 │ 3000                   │ 100         │ 64              │
│ master                  │  0.60405 │  0.9608 │ 3000                   │ 100         │ 64              │
│ ├── d384680 [exp-bc055] │  0.51799 │ 0.92333 │ 500                    │ 100         │ 64              │
│ ├── 6b338f8 [exp-3315b] │  0.58589 │   0.945 │ 2000                   │ 100         │ 64              │
│ └── d7fdde2 [exp-1b262] │  0.56447 │ 0.94713 │ 1500                   │ 100         │ 64              │
└─────────────────────────┴──────────┴─────────┴────────────────────────┴─────────────┴─────────────────┘
```

Sort experiments by the `auc` metric, in descending order:

```dvc

> $ dvc exp show --only-changed --sort-by=roc_auc --sort-order desc

Reviewer 3: Seems unrelated? And there's already a sorting example after this.

Author: In the original examples there were also 2 examples for `--sort-by`. 

Only change I did here is to replace `--sort-order asc` with `--sort-order desc` because `--sort-order asc` has the same behavior as not passing the option at all, and could give the impression that you need to pass the value in order to actually sort.

┡━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ workspace               │ -            │  0.60405 │  0.9608 │ 3000                   │ 2                │ 20170428   │ 100         │ 64              │
│ master                  │ May 29, 2021 │  0.60405 │  0.9608 │ 3000                   │ 2                │ 20170428   │ 100         │ 64              │
│ ├── 6b338f8 [exp-3315b] │ 08:03 PM     │  0.58589 │   0.945 │ 2000                   │ 2                │ 20170428   │ 100         │ 64              │
│ ├── d384680 [exp-bc055] │ 08:03 PM     │  0.51799 │ 0.92333 │ 500                    │ 2                │ 20170428   │ 100         │ 64              │
│ └── d7fdde2 [exp-1b262] │ 08:03 PM     │  0.56447 │ 0.94713 │ 1500                   │ 2                │ 20170428   │ 100         │ 64              │
└─────────────────────────┴──────────────┴──────────┴─────────┴────────────────────────┴──────────────────┴────────────┴─────────────┴─────────────────┘
```

> You can use `|` in order to remove a list of columns:

Author: ```suggestion
You can use [regex][regex] to match columns. For example, to remove multiple columns:
```

`--keep` has the last word:

```dvc
$ dvc exp show --only-changed --drop Created --keep 'train.(?!seed)'
```

```dvctable
┏━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ neutral:**Experiment**              ┃ metric:**avg_prec** ┃ metric:**roc_auc** ┃ param:**featurize.max_features** ┃ param:**train.n_est** ┃ param:**train.min_split** ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ workspace               │  0.60405 │  0.9608 │ 3000                   │ 100         │ 64              │
│ master                  │  0.60405 │  0.9608 │ 3000                   │ 100         │ 64              │
│ ├── d384680 [exp-bc055] │  0.51799 │ 0.92333 │ 500                    │ 100         │ 64              │
│ ├── 6b338f8 [exp-3315b] │  0.58589 │   0.945 │ 2000                   │ 100         │ 64              │
│ └── d7fdde2 [exp-1b262] │  0.56447 │ 0.94713 │ 1500                   │ 100         │ 64              │
└─────────────────────────┴──────────┴─────────┴────────────────────────┴─────────────┴─────────────────┘
```

> Sort experiments by the `auc` metric, in descending order:

Reviewer: ```suggestion
Sort experiments by the `roc_auc` metric, in descending order:
```

```dvc
$ dvc exp show --only-changed --sort-by=roc_auc --sort-order desc
```

```dvctable
┏━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ neutral:**Experiment**              ┃ neutral:**Created**      ┃ metric:**avg_prec** ┃ metric:**roc_auc** ┃ param:**featurize.max_features** ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│ workspace               │ -            │  0.60405 │  0.9608 │ 3000                   │
│ master                  │ May 29, 2021 │  0.60405 │  0.9608 │ 3000                   │
│ ├── d7fdde2 [exp-1b262] │ 08:03 PM     │  0.56447 │ 0.94713 │ 1500                   │
│ ├── 6b338f8 [exp-3315b] │ 08:03 PM     │  0.58589 │   0.945 │ 2000                   │
│ └── d384680 [exp-bc055] │ 08:03 PM     │  0.51799 │ 0.92333 │ 500                    │
└─────────────────────────┴──────────────┴──────────┴─────────┴────────────────────────┘
```

To see all experiments throughout the Git history:

```dvc

> $ dvc exp show --all-commits --only-changed --sort-by=auc

Reviewer: ```suggestion
$ dvc exp show --all-commits --only-changed --sort-by=roc_auc
```

📖 See [Metrics, Parameters, and Plots](/doc/start/metrics-parameters-plots) for
an introduction to parameters, metrics, plots.

> [regex]: https://en.wikipedia.org/wiki/Regular_expression

Reviewer 2: I think reading this page makes people more confused than searching for "regexp." Could you replace it with a simpler page, like https://towardsdatascience.com/an-introduction-to-regular-expressions-5dd762afc5e4 ?

Author: Replaced with https://regexone.com/
